File: BootCampOOP/Unit/Unit.cpp
#include "Unit.h"

Unit::Unit(const std::string& nameIn, int hpIn, int dmgIn){
    
    name = nameIn;
    hitPointsLimit = hpIn;
    damage = dmgIn;
    hitPoints = hitPointsLimit;
    
}

Unit::~Unit(){
    
}

void Unit::ensureIsAlive() {
    
   if ( hitPoints == 0 ) {
       throw UnitIsDead();
   }
}

int Unit::getDamage() const{
    return damage;
    
}

int Unit::getHitPoints() const{
    return hitPoints;
    
}

int Unit::getHitPointsLimit() const{
    return hitPointsLimit;
    
}

const std::string& Unit::getName() const{
    return name;
    
}

void Unit::takeDamage(int dmg){

    if (dmg < hitPoints){
        hitPoints -= dmg;
    }
    else {
        hitPoints = 0;
    }
}

void Unit::addHitPoints(int hp){
    
    ensureIsAlive();
    
    if (hitPoints + hp > hitPointsLimit){
        hitPoints = hitPointsLimit;
    }
    else {
        hitPoints += hp;
    }    
    
}

void Unit::attack(Unit& enemy){
    
    ensureIsAlive();
    enemy.ensureIsAlive();
    
    enemy.takeDamage(damage);
    
    if (enemy.getHitPoints() > 0){
        enemy.counterAttack(*this);
    }
    ensureIsAlive();
    enemy.ensureIsAlive();
}

void Unit::counterAttack(Unit& enemy){
    enemy.takeDamage(damage / 2);
}

std::ostream& operator<<(std::ostream& out, const Unit& unit){
    out << "Name: " << unit.getName() << std::endl;
    out << "Кол-во HP: " << unit.getHitPointsLimit() << std::endl;
    out << "Сила урона: " << unit.getDamage() << std::endl;
    out << "Кол-во HP после атаки: " << unit.getHitPoints() << std::endl;
}

File: BootCamp Java/Unit/Unit.java
public class Unit {
    private int damage;
    private int hitPoints;
    private int hitPointsLimit;
    private String name;
    
    public Unit(String name, int hp, int dmg) {
        this.damage = dmg;
        this.hitPointsLimit = hp;
        this.hitPoints = hitPointsLimit;
        this.name = name;
    }
    
    public int getDamage() {
        return this.damage;
    }
    public int getHitPoints() {
        return this.hitPoints;
    }

    public int getHitPointsLimit() {
        return this.hitPointsLimit;
    }
    public String getName() {
        return this.name;
    }

    private void ensureIsAlive() throws UnitIsDeadException {
        if ( this.hitPoints<1 ) {
            throw new UnitIsDeadException(
            );
        }
    }

    public void addHitPoints(int hp) throws UnitIsDeadException {
        ensureIsAlive();
        this.hitPoints += hp;
        if ( this.hitPoints > hitPointsLimit ) {
            this.hitPoints = hitPointsLimit;
        }
    }

    public void takeDamage(int dmg) throws UnitIsDeadException {
        ensureIsAlive();
        this.hitPoints -= dmg;
        if ( this.hitPoints <=0 ) {
            this.hitPoints = 0;
        }
    }

    public void attack(Unit enemy) throws UnitIsDeadException{
        ensureIsAlive();
        enemy.takeDamage(damage);
        counterAttack(enemy);
    }

    public void counterAttack(Unit enemy) throws UnitIsDeadException{
        enemy.ensureIsAlive();
        takeDamage(enemy.damage/2);
    }

    public String toString() {
        return "Unit name is:"+getName()+'\n'+"Unit damage:"+getDamage()+'\n'+"Unit hit point limit:"+getHitPointsLimit()+'\n'+"Unit current hit point:"+getHitPoints()+'\n';
    }
}